reconcile/utils/pagerduty_api.py

```diff
@@ -1,40 +1,77 @@
-import datetime
+from datetime import datetime as dt
+import logging
+from typing import Callable
+from typing import Iterable
+from typing import Optional
+from typing import Protocol
 
-import requests
 import pypd
+import requests
+from pydantic import BaseModel
+from sretoolbox.utils import retry
 
 from reconcile.utils.secret_reader import SecretReader
+from reconcile.utils.secret_reader import SupportsSecret
+
+
+class PagerDutyTargetException(Exception):
+    ...
+
+
+class SupportsPagerDutyInstance(Protocol):
+    name: str
+
+    @property
+    def token(self) -> SupportsSecret:
+        ...
+
+
+class SupportsUser(Protocol):
+    org_username: str
+    pagerduty_username: Optional[str]
+
+
+class SupportsPagerDutyTarget(Protocol):
+    name: str
+    instance: SupportsPagerDutyInstance
+    escalation_policy_id: Optional[str]
+    schedule_id: Optional[str]
+
+
+class PagerDutyConfig(BaseModel):
+    name: str
+    token: str
 
 
 class PagerDutyApi:
     """Wrapper around PagerDuty API calls"""
 
-    def __init__(self, token, init_users=True, settings=None):
-        secret_reader = SecretReader(settings=settings)
-        pd_api_key = secret_reader.read(token)
-        pypd.api_key = pd_api_key
+    def __init__(self, token: str, init_users: bool = True) -> None:
+        pypd.api_key = token
         if init_users:
             self.init_users()
         else:
-            self.users = []
+            self.users: list[pypd.User] = []
 
-    def init_users(self):
+    def init_users(self) -> None:
         self.users = pypd.User.find()
 
-    def get_pagerduty_users(self, resource_type, resource_id):
-        now = datetime.datetime.utcnow()
+    def get_pagerduty_users(
+        self, resource_type: str, resource_id: str
+    ) -> list[pypd.User]:
+        now = dt.utcnow()
 
         try:
             if resource_type == "schedule":
                 users = self.get_schedule_users(resource_id, now)
             elif resource_type == "escalationPolicy":
                 users = self.get_escalation_policy_users(resource_id, now)
         except requests.exceptions.HTTPError:
-            return None
+            return []
```

Comment:
i'm aware that this kind of errorhandling was already there before the refactor and that the calling code deals with this, but this approach makes me uneasy.

no need to fix in this context because you did not change anything in behaviour

Reply:
I've added type hints, and therefore I need to decide on `list[...]` vs. `Optional[list[...]]`. The `slack-usergroups` integration, which uses `get_pagerduty_users`, will be refactored later and will benefit :)

Reply:
i meant that this function hides the fact that an error occured and that `None` or `[]` might not be the actual state on pagerduty

Reply:
Let me add a log message

Reply:
works for me for now 👍

Reply:
[2946115](https://github.com/app-sre/qontract-reconcile/pull/3002/commits/29461151d9053890e67ba21ecc05e1e6fb6da813) ❤️